File: ResourceManager/Services/PermissionDataService.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Design;
using ResourceManager.Models;
using ResourceManager.Services.Abstractions;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Threading.Tasks;

namespace ResourceManager.Services
{
    public class PermissionDataService : GenericDataService<Permission>, IPermissionDataService
    {
        public PermissionDataService(IDesignTimeDbContextFactory<ResourceManagerDBContext> contextFactory) : base(contextFactory)
        {
        }
        public override async Task<IList<Permission>> GetAll()
        {
            using (ResourceManagerDBContext db = dbFactory.CreateDbContext(null!))
            {
                IList<Permission> entities = await db.Permissions.OrderBy(x => x.Name).ToListAsync();
                return entities;
            }
        }
        public async Task<IList<string>> GetPermissionForRoles(IList<string> roles)
        {
            using (ResourceManagerDBContext db = dbFactory.CreateDbContext(null!))
            {
                var results = await (from p in db.Permissions
                                     from r in db.RolePermissions.Where(x=>roles.Contains(x.Role.Name) && x.PermissionID == p.ID)
                                     select p.Name).Distinct().ToListAsync();
                return results;
            }
            
        }
        public async Task<IList<Permission>> GetPermissionObjectForRoleID(int id)
        {
            using (ResourceManagerDBContext db = dbFactory.CreateDbContext(null!))
            {
                var results = await (from p in db.Permissions
                                     from r in db.RolePermissions.Where(x => x.RoleID==id && x.PermissionID == p.ID)
                                     select p)
                                     .Distinct()
                                     .OrderBy(x=>x.Name)
                                     .ToListAsync();
                return results;
            }
        }
    }
}

File: ResourceManager/Services/PermissionDataService.vb
Imports System.Data
Imports Microsoft.EntityFrameworkCore
Imports Microsoft.EntityFrameworkCore.Design
Imports ResourceManager.Models
Imports ResourceManager.Services.Abstractions

Namespace ResourceManager.Services
    Public Class PermissionDataService
        Inherits GenericDataService(Of Permission)
        Implements IPermissionDataService
        Public Sub New(ByVal contextFactory As IDesignTimeDbContextFactory(Of ResourceManagerDBContext))
            MyBase.New(contextFactory)
        End Sub
        Public Overrides Async Function GetAll() As Task(Of IList(Of Permission)) Implements IDataService(Of Permission).GetAll
            Dim db = MyBase.dbFactory.CreateDbContext(Nothing)
            Using db
                Dim entities = Await db.Permissions.OrderBy(Function(x) x.Name).ToListAsync()
                Return entities
            End Using
        End Function
        Public Async Function GetPermissionForRoles(ByVal roles As IList(Of String)) As Task(Of IList(Of String)) Implements IPermissionDataService.GetPermissionForRoles
            Dim db = MyBase.dbFactory.CreateDbContext(Nothing)
            Using db
                Dim results = Await (From p In db.Permissions
                                     Join r In db.RolePermissions
                                         On r.PermissionID Equals p.ID
                                     Where roles.Contains(r.Role.Name)
                                     Select p.Name).Distinct().ToListAsync()
                Return results
            End Using
        End Function
        Public Async Function GetPermissionObjectForRoleID(ByVal id As Integer) As Task(Of IList(Of Permission)) Implements IPermissionDataService.GetPermissionObjectForRoleID
            Dim db = MyBase.dbFactory.CreateDbContext(Nothing)
            Using db
                Dim results = Await (From p In db.Permissions
                                     Join r In db.RolePermissions
                                         On p.ID Equals r.PermissionID
                                     Where r.RoleID = id
                                     Order By p.Name
                                     Select p).Distinct().ToListAsync
                Return results
            End Using
        End Function
    End Class
End Namespace
